### ISSUE TITLE
Window.Activate does not activate and bring window to foreground in WinUI3 if window is in background but not minimized

### ISSUE THREAD

**Issue by Balkoth** [microsoft/microsoft-ui-xaml#7595]
### Describe the bug

If `Activate` is called on a `Window` which is in the background of other windows, that window is not activated and brought to the foreground. If the window is minimized it works correctly.

### Steps to reproduce the bug

Call `Activate` on a `Window` when it is in the backround of other windows.

### NuGet package version

WinUI 3 - Windows App SDK 1.1.4

### Windows app type

- [ ] UWP
- [X] Win32

### Device form factor

Desktop

### Windows version

Windows 10 (21H2): Build 19044

**Comment by msalandro** [comment]
I ran into this issue when making my WinUI3 app single-instanced. Any idea when the bug could be resolved?

**Comment by Balkoth** [comment]
Exactly where the bug hit me too. Had to call SetForegroundWindow. Please fix this.

**Comment by pratikone** [comment]
looking at it

**Comment by pratikone** [comment]
I just tried in the latest version of windows app sdk 1.2 and not able to reproduce  it. I created a bunch of windows, stacked them one behind the other, called `.Activate()` on one of them and it showed up. 

Can you verify that after updating to winappsdk 1.2, it goes away ? Feel free to re-open this bug if it still persists.

**Comment by Balkoth** [comment]
This is not fixed in 1.2! Running code like this

```cs
// Wire up the activated event handler.
keyInstance.Activated += async (sender, e) =>
{
  // When activated, activate the main window and bring it to the foreground.
  await _mainWindow.DispatcherQueue.EnqueueAsync(() => _mainWindow.Activate());
  //NativeMethods.SetForegroundWindow(_mainWindow.GetWindowHandle());
};
```

outside the debugger, does not bring the window to the foreground. Only once the native call is uncommented the window is brought to the foreground!

**Comment by pratikone** [comment]
thanks for verifying it. i will try this out quickly.

**Comment by pratikone** [comment]
It looks like you are using Window community toolkit for dispatcher queue. Could you share a sample project which I can compile and build locally to investigate the issue?

**Comment by Balkoth** [comment]
Is there a sample available for the Windows App SDK dispatcher queue? The link you posted puts a big question mark over my head...

**Comment by pratikone** [comment]
WinUI Gallery source code has some parts which uses this inbuilt dispatcher queue :   https://github.com/microsoft/WinUI-Gallery/blob/353c1a0dab0aec9485179ad2a8b76cab95b7c8b9/WinUIGallery/TabViewPages/TabViewWindowingSamplePage.xaml.cs#L179

See if it helps.

**Comment by Balkoth** [comment]
I will try it tomorrow and report back.

**Comment by Balkoth** [comment]
Using the Windows App SDK dispatcher queue does also not work...

```cs
// Wire up the activated event handler.
keyInstance.Activated += async (sender, e) =>
{
  var taskCompletionSource = new TaskCompletionSource();
  _mainWindow.DispatcherQueue.TryEnqueue(
      Microsoft.UI.Dispatching.DispatcherQueuePriority.Normal,
      new Microsoft.UI.Dispatching.DispatcherQueueHandler(() =>
      {
        _mainWindow.Activate();
        taskCompletionSource.SetResult();
      }));

  await taskCompletionSource.Task;
};
```

**Comment by Balkoth** [comment]
Here is the sample project: [WindowActivateTest.zip](https://github.com/microsoft/microsoft-ui-xaml/files/10220167/WindowActivateTest.zip)

**Comment by pratikone** [comment]
thanks for confirming. i will try out this test app.

**Comment by pratikone** [comment]
Thanks for sample app. I debugged it. Turns out AppInstance.Activated event is not firing and that's why your window is not coming forward. 

If you hook this code to any other event's firing like button clicked, it starts working.
![image](https://user-images.githubusercontent.com/115935/208986237-c7f0ffe8-6f0f-4dc4-9643-84463195500c.png).

So, you can debug AppInstance code to see why it is not working.

**Comment by Balkoth** [comment]
WTF are you talking about. Of course the event is called, if you put in the native call to SetForegroundWindow the window is activated an put to the foreground. Clearly you dont seem to be qualified to analyse this issue, please bring a colleage with more experience.

**Comment by Balkoth** [comment]
So will this be reopened and investigated properly or do i have to create a new issue?

**Comment by pratikone** [comment]
My apologies. I was not launching the second instance. Now I am able to get the repro done. 
We are investigating this and I will keep the issue open until you validate the fix.

**Comment by Jay-o-Way** [comment]
Still investigating?

**Comment by Jay-o-Way** [comment]
Question: Why/when use this methiod over `.Show()` ?

**Comment by abbasghomi** [comment]
This worked for me (I need app to be single instanced too and if user tried to run app again, already running instance's window get activated)

This worked for me

```
[DllImport("user32.dll")]
public static extern bool ShowWindow(IntPtr hWnd, int nCmdShow);

[DllImport("user32.dll")]
public static extern bool SetForegroundWindow(IntPtr hWnd);

public enum ShowWindowCommands : int
{
	Hide = 0,
	ShowNormal = 1,
	ShowMinimized = 2,
	ShowMaximized = 3,
	Maximize = 3,
	ShowNormalNoActivate = 4,
	Show = 5,
	Minimize = 6,
	ShowMinNoActivate = 7,
	ShowNoActivate = 8,
	Restore = 9,
	ShowDefault = 10,
	ForceMinimized = 11
}
```

in app.Xaml.cs
        
```
protected override async void OnLaunched(LaunchActivatedEventArgs args)
{
	// Get the activation args
	var appArgs = AppInstance.GetCurrent().GetActivatedEventArgs();
	
	// Get or register the main instance
	var mainInstance = AppInstance.FindOrRegisterForKey("Bartlett.CRM.App");
	// If the main instance isn't this current instance
	
	if (!mainInstance.IsCurrent)
		{
			var activatedEventArgs = AppInstance.GetCurrent().GetActivatedEventArgs();
			await mainInstance.RedirectActivationToAsync(activatedEventArgs);
			Process.GetCurrentProcess().Kill();
			return;
		}
	// Otherwise, register for activation redirection
	AppInstance.GetCurrent().Activated += AppActivated;
}

private void AppActivated(object sender, AppActivationArguments e)
{
	IntPtr hWnd = WindowNative.GetWindowHandle(_mainWindow);
	//this line causing a delay in reactivating main window, it needed to restore window if it was minimized
	ShowWindow(hWnd, (int)NativeMethods.ShowWindowCommands.ShowNormal);
	//this method will bring main window to front
	SetForegroundWindow(hWnd);
}        
```

**Comment by Balkoth** [comment]
This is still an issue in Windows App SDK 1.3.2 (1.3.230602002)

**Comment by AnalogFeelings** [comment]
Seems to still be a problem in WASDK 1.4, is there any ETA for this?

**Comment by Balkoth** [comment]
Does the explorer team using WinUi3 also just work around bugs like this? I would get really mad if i was on that team and long-standing bugs like this are still open.

**Comment by Jimex** [comment]
> IntPtr hWnd = WindowNative.GetWindowHandle(_mainWindow);
> 	//this line causing a delay in reactivating main window, it needed to restore window if it was minimized
> 	ShowWindow(hWnd, (int)NativeMethods.ShowWindowCommands.ShowNormal);
> 	//this method will bring main window to front
> 	SetForegroundWindow(hWnd);

Worked like a champ, thank you.

**Comment by tipa** [comment]
I was hitting the same issue today as well. My goal is to activate/foreground the apps single-instanced window after the user clicks a toast notification of the same app (a pretty common scenario I would say). The `SetForegroundWindow` workaround posted above does not work if the debugger isn't attached due to the reasons [here](https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-setforegroundwindow#remarks).

A similar behavior can be observed when using `Windows.System.Launcher.LaunchUriAsync`. Normally this would launch *and foreground* your web browser if you call it, e.g. after clicking a button in the main window. However if you call it in `Program.Main` it launches your browser, but doesn't foreground it.

**Comment by castorix** [comment]
> I was hitting the same issue today as well. My goal is to activate/foreground the apps single-instanced window after the user clicks a toast notification of the same app (a pretty common scenario I would say). The `SetForegroundWindow` workaround posted above does not work if the debugger isn't attached due to the reasons [here](https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-setforegroundwindow#remarks).
> 

An API which generally works is [SwitchToThisWindow](https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-switchtothiswindow)
(they say "It may be altered...", but it is used by Task Manager for ages...)

**Comment by tipa** [comment]
@castorix sadly, this also doesn't work for me...

**Comment by castorix** [comment]
> @castorix sadly, this also doesn't work for me...

I converted an old C++ function into C#, mixing all APIs.
If it does not work for you, I have no more idea...

        private void SwitchToWindow(IntPtr hWnd)
        {
            int nLockTimeOut = 0;
            IntPtr hCurrWnd = GetForegroundWindow();
            int nPID = 0;
            uint nThisTID = GetCurrentThreadId(), nCurrTID = GetWindowThreadProcessId(hCurrWnd, out nPID);
            if (nThisTID != nCurrTID)
            {
                AttachThreadInput(nThisTID, nCurrTID, true);
                SystemParametersInfo(SPI_GETFOREGROUNDLOCKTIMEOUT, 0, ref nLockTimeOut, 0);
                int nNewLockTimeOut = 0;
                SystemParametersInfo(SPI_SETFOREGROUNDLOCKTIMEOUT, 0, ref nNewLockTimeOut, SPIF_SENDWININICHANGE | SPIF_UPDATEINIFILE);
                AllowSetForegroundWindow(-1);
            }
            if (IsIconic(hWnd))
                ShowWindow(hWnd, SW_RESTORE);
            IntPtr hWndLastActivePopup = GetLastActivePopup(hWnd);
            SwitchToThisWindow(hWndLastActivePopup, true);           
            SetForegroundWindow(hWnd);
            if (nThisTID != nCurrTID)
            {
                SystemParametersInfo(SPI_SETFOREGROUNDLOCKTIMEOUT, 0, ref nLockTimeOut, SPIF_SENDWININICHANGE | SPIF_UPDATEINIFILE);
                AttachThreadInput(nThisTID, nCurrTID, false);
            }
        }

with :

        [DllImport("User32.dll", SetLastError = true, CharSet = CharSet.Unicode)]
        //public static extern bool SystemParametersInfo(uint uiAction, uint uiParam, IntPtr pvParam, uint fWinIni);
        public static extern bool SystemParametersInfo(uint uiAction, uint uiParam, [In, Out] ref int pvParam, uint fWinIni);

        public const int SPIF_UPDATEINIFILE = 0x0001;
        public const int SPIF_SENDWININICHANGE = 0x0002;

        public const int SPI_GETFOREGROUNDLOCKTIMEOUT = 0x2000;
        public const int SPI_SETFOREGROUNDLOCKTIMEOUT = 0x2001;

        [DllImport("User32.dll", SetLastError = true, CharSet = CharSet.Unicode)]
        public static extern IntPtr GetForegroundWindow();

        [DllImport("Kernel32.dll", SetLastError = true, CharSet = CharSet.Unicode)]
        public static extern uint GetCurrentThreadId();

        [DllImport("User32.dll", SetLastError = true, CharSet = CharSet.Unicode)]
        public static extern uint GetWindowThreadProcessId(IntPtr hWnd, out int lpdwProcessId);

        [DllImport("User32.dll", SetLastError = true, CharSet = CharSet.Unicode)]
        public static extern bool AttachThreadInput(uint idAttach, uint idAttachTo, bool fAttach);

        [DllImport("User32.dll", SetLastError = true, CharSet = CharSet.Unicode)]
        public static extern bool AllowSetForegroundWindow(int dwProcessId);

        [DllImport("User32.dll", SetLastError = true, CharSet = CharSet.Unicode)]
        public static extern bool ShowWindow(IntPtr hWnd, int nCmdShow);

        public const int SW_HIDE = 0;
        public const int SW_SHOWNORMAL = 1;
        public const int SW_SHOWMINIMIZED = 2;
        public const int SW_SHOWMAXIMIZED = 3;
        public const int SW_RESTORE = 9;

        [DllImport("User32.dll", SetLastError = true, CharSet = CharSet.Unicode)]
        public static extern bool IsIconic(IntPtr hWnd);

        [DllImport("User32.dll", SetLastError = true, CharSet = CharSet.Unicode)]
        public static extern IntPtr GetLastActivePopup(IntPtr hWnd);

        [DllImport("User32.dll", SetLastError = true, CharSet = CharSet.Unicode)]
        public static extern bool SwitchToThisWindow(IntPtr hWnd, bool fAltTab);

        [DllImport("User32.dll", SetLastError = true, CharSet = CharSet.Unicode)]
        public static extern bool SetForegroundWindow(IntPtr hWnd);

**Comment by ritesh3103** [comment]
I did this as a temporary solution.

In App.xaml.cs

```
protected async override void OnLaunched(Microsoft.UI.Xaml.LaunchActivatedEventArgs args)
{
	var mainInstance = Microsoft.Windows.AppLifecycle.AppInstance.FindOrRegisterForKey("main");
	var activatedEventArgs = Microsoft.Windows.AppLifecycle.AppInstance.GetCurrent().GetActivatedEventArgs();

	if (!mainInstance.IsCurrent)
	{
   
		await mainInstance.RedirectActivationToAsync(activatedEventArgs);
		System.Diagnostics.Process.GetCurrentProcess().Kill();
		return;
	}
	m_window = new MainWindow();
	m_window.Activate();
	mainInstance.Activated += MainInstance_Activated;
}
```


```
private void MainInstance_Activated(object sender, Microsoft.Windows.AppLifecycle.AppActivationArguments e)
{
	General.WinPresenter.Minimize();
	General.WinPresenter.Restore();
}
```

In MainWindow:

```
public MainWindow()
{
    this.InitializeComponent();
    General.WndHnd = WinRT.Interop.WindowNative.GetWindowHandle(this);
}
```

```
private void WinLoaded()
{
	AppWindow CurWin;
	WindowId WinID = Win32Interop.GetWindowIdFromWindow(General.WndHnd);
	CurWin = AppWindow.GetFromWindowId(WinID);

	General.WinPresenter = CurWin.Presenter as OverlappedPresenter;
}
```
common class in the project

```
internal class General
{
	public static IntPtr WndHnd;
	public static Microsoft.UI.Windowing.OverlappedPresenter WinPresenter;
}
```

**Comment by jingliancui** [comment]
hi Team, is there any update on this issue? Thanks.

**Comment by Balkoth** [comment]
Will this issue anytime soon be fixed?

**Comment by pratikone** [comment]
Thanks. It is in the list of bugs we have for fixing. Looking into it. Will update once I have something

**Comment by pratikone** [comment]
Some update : i have found the root issue and the fix. Now remains is the internal discussion over the fix.
Here is some explanation of why it happpens : 

This is due to the fact Window.Activate() internally calls SetActiveWindow since its inception ([source](https://github.com/microsoft/microsoft-ui-xaml/blob/winui3/release/1.5-stable/dxaml/xcp/dxaml/lib/DesktopWindowImpl.cpp#L416)) .  
WPF on the other hand, calls `SetForegroundWindow()`.  Hence, this issue is only for winui and not WPF.

now, WinUI could start calling SetForegroundWindow too but it would be a breaking change. We are currently assessing the impact on backward compatibility and then we will do the fix.

**Comment by Balkoth** [comment]
IMHO this is exactly how `Activate` should always have behaved and is documented:
<img width="670" alt="image" src="https://github.com/microsoft/microsoft-ui-xaml/assets/80055480/b2229c7b-1732-4bd3-9b36-ae5aa343fbc5">

**Comment by WillyWoe** [comment]
How about creating a new method that calls SetForegroundWindow(), like Window.BringToForeground() to avoid the breaking change in Window.Activate()

**Comment by k1mmm** [comment]
Any progress here? =)

**Comment by garrettpauls** [comment]
Is there any plan to address this? We're nearing six months since a fix was identified for this and there doesn't seem to have been any movement since then. Could we at least get the [documentation](https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.ui.xaml.window.activate) updated with a remark describing the problem and linking back here so others don't lose time trying to figure out why windows aren't coming to the foreground when activated?

**Comment by Balkoth** [comment]
What is happening now to this issue? Is there any plan going ahead?

**Comment by applefanbois** [comment]
Is this retarded Microslop company ever gonna make Windows work on Windows OS?

Contrary to  macOS, iOS, Android and Linux, Microsoft does not bring the app to top/front when it receive a file or a url.
And there is no working ways of forcing the app window on top so the user can see that it received a file.

This problem is now 3 years old and they don't even bother to fix this most basic feature that is essential for any app. When people complain to us that things are not working on the windows version of our app we always tell them that Microsoft has abandoned Windows 10 years ago and they should go to a supported modern operating system like macOS or Android/ChromeOS.

I am dreaming of the day when we no longer forced to suffer to make app for windows. Maybe we have another 5 - 10 years to wait?

Windows is an abandoned platform, they only care about ads and steeling their customers data to feed their AI slop. Developing for windows is  Chinese water torture.

**Comment by applefanbois** [comment]
If I try to minimize the window first then wait 5 seconds then call appWindow.Show() it still does not work.


But @DeshaunZ  solution works.
Just copy the code below directly where you use it.

```
            [DllImport("user32.dll")]
            static extern bool SetForegroundWindow(IntPtr hWnd);
            SetForegroundWindow(hWnd);
```

I will just add that with broken native SDK/API like that, no wonder younger generation are going full WEB.

**Comment by catmanjan** [comment]
Bump

**Comment by applefanbois** [comment]
MS won't fix anything, because Windows has only 5 years left before everyone has moved to superior and simpler Linux.

Call this manually instead :

    public void bringToTop()
    { 
        Window? window = /*ask an ai where to get the window*/

        IntPtr hWnd = WindowNative.GetWindowHandle(window);

        [DllImport("user32.dll")]
        static extern bool SetForegroundWindow(IntPtr hWnd);
        SetForegroundWindow(hWnd);
    }

**Comment by catmanjan** [comment]
I am already but it bothers me that its necessary

Isn’t this project open source now anyway?

**Comment by applefanbois** [comment]
Everything at MS is vibe coded by Copislop+ and by the 10 000 "new temporary Americans" that have replaced original employees. The 2 people working on the code probably left to contribute to useful things instead (like Linux things).

I don't take any new Windows contracts. I only do macOS and Linux jobs now. More work for you lol.

Good luck fren.

**Comment by elibroftw** [comment]
> MS won't fix anything, because Windows has only 5 years left before everyone has moved to superior and simpler Linux.
> 
> Call this manually instead :
> 
> ```
> public void bringToTop()
> { 
>     Window? window = /*ask an ai where to get the window*/
> 
>     IntPtr hWnd = WindowNative.GetWindowHandle(window);
> 
>     [DllImport("user32.dll")]
>     static extern bool SetForegroundWindow(IntPtr hWnd);
>     SetForegroundWindow(hWnd);
> }
> ```

If you use [CsWin32](https://microsoft.github.io/CsWin32/docs/getting-started.html), you can simply maintain a list of methods in NativeMethods.txt rather than using DllImport.

```cs
// App.xaml.cs
    MainWindow? _window;

    internal void ShowMainWindow()
    {
        Log.Info("activating main window");
        PInvoke.ShowWindow(_window.GetWindowHandle(), SHOW_WINDOW_CMD.SW_NORMAL);
        PInvoke.SetForegroundWindow(_window.GetWindowHandle());
        Log.Info("activated main window");
    }

// MainWindow.xaml.cs
    internal HWND GetWindowHandle()
    {
        return (HWND)WinRT.Interop.WindowNative.GetWindowHandle(this);
    }
```

**Comment by bogdan-patraucean** [comment]
I can confirm that using `SetForegroundWindow` API is fixing the issue. I don't understand why the WinUI 3 is not able to replicate the same behaviour as a native win32 API. I discovered this issue while using CTRL + Shift + Click to open an app as administrator: if you don't release CTRL + Shift keys fast enough, the window will not be set in the foreground.

**Comment by JochemPalmsens** [comment]
@JesseCol how was this done/completed? Is it fixed?

**Comment by Balkoth** [comment]
Seems nothing was fixed with `Activate` and a new method was invented which just calls `SetForegroundWindow`. Sorry but this is not what was asked or reported here. If the window is minimized `Activate` actually brings the window to the foreground. Only when it is in the background of other windows it does not.

**Comment by JesseCol** [comment]
> Seems nothing was fixed with `Activate` and a new method was invented which just calls `SetForegroundWindow`. Sorry but this is not what was asked or reported here. If the window is minimized `Activate` actually brings the window to the foreground. Only when it is in the background of other windows it does not.

Correct!  Yes, we added an experimental API Window.TrySetForeground.  I don't want to change the behavior of Window.Activate here, since it's the well-known way to show a WinUI Window for years, and does the basic win32 ShowWindow call to get it visible.  (maybe Show() would be a better name for this function).

But I suspect TrySetForeground doesn't actually solve your problem anyway, since Windows only honors this request in certain situations.  See remarks section in [this doc](https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-setforegroundwindow).  So, I've reactivated this issue.

**Comment by JesseCol** [comment]
>I can confirm that using SetForegroundWindow API is fixing the issue

I noticed this comment above, so maybe there's more benefit to exposing this than I'm realizing.  Can y'all help me understand your scenarios better?

If you just call SetForegroundWindow "out of the blue" when some other app has foreground, Windows won't honor it, it will just flash your taskbar icon, which isn't something it sounds like folks on this thread want (?)

If you're making a single-instance app, typically when the 2nd instance of the app is launched, it redirects the user to the 1st instance and closes itself.  In this case, the 2nd instance can call [AllowSetForegroundWindow](https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-allowsetforegroundwindow) to allow the 1st instance to grab foreground.

Do you have other specific scenarios related to this issue?  Thanks!

**Comment by Felicity-R** [comment]
> > I can confirm that using SetForegroundWindow API is fixing the issue
> 
> I noticed this comment above, so maybe there's more benefit to exposing this than I'm realizing. Can y'all help me understand your scenarios better?
> 
> If you just call SetForegroundWindow "out of the blue" when some other app has foreground, Windows won't honor it, it will just flash your taskbar icon, which isn't something it sounds like folks on this thread want (?)
> 
> If you're making a single-instance app, typically when the 2nd instance of the app is launched, it redirects the user to the 1st instance and closes itself. In this case, the 2nd instance can call [AllowSetForegroundWindow](https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-allowsetforegroundwindow) to allow the 1st instance to grab foreground.
> 
> Do you have other specific scenarios related to this issue? Thanks!

The specific scenario for my team is that we have set up deep links (like my-app://?someargs) for our app. When a customer uses one of these links, they expect the app to appear in the foreground if it was already running. 
We have since added the SetForegroundWindow bit so it works now, but in our older version of the app (back in the Xamarin Forms days) that wasn't necessary, so this feels like a regression.

Edit: Oh, and we are using a single-instanced app too.

**Comment by JesseCol** [comment]
> > > I can confirm that using SetForegroundWindow API is fixing the issue
> > 
> > 
> > I noticed this comment above, so maybe there's more benefit to exposing this than I'm realizing. Can y'all help me understand your scenarios better?
> > If you just call SetForegroundWindow "out of the blue" when some other app has foreground, Windows won't honor it, it will just flash your taskbar icon, which isn't something it sounds like folks on this thread want (?)
> > If you're making a single-instance app, typically when the 2nd instance of the app is launched, it redirects the user to the 1st instance and closes itself. In this case, the 2nd instance can call [AllowSetForegroundWindow](https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-allowsetforegroundwindow) to allow the 1st instance to grab foreground.
> > Do you have other specific scenarios related to this issue? Thanks!
> 
> The specific scenario for my team is that we have set up deep links (like my-app://?someargs) for our app. When a customer uses one of these links, they expect the app to appear in the foreground if it was already running. We have since added the SetForegroundWindow bit so it works now, but in our older version of the app (back in the Xamarin Forms days) that wasn't necessary, so this feels like a regression.
> 
> Edit: Oh, and we are using a single-instanced app too.

Thanks!  That's a good one.  Another similar one could be activating your window based on a toast notification.

Do you remember the API that made things "just work" before in Xamarin?  It doesn't seem right to me to make Window.Activate also call SetForegroundWindow, but it could be I'm missing something.

**Comment by Felicity-R** [comment]
> Do you remember the API that made things "just work" before in Xamarin? It doesn't seem right to me to make Window.Activate also call SetForegroundWindow, but it could be I'm missing something.

It would have been UWP instead of WinUI3. We just called Window.Current.Activate(), from the Windows.UI.Xaml.Application.OnActivated(IActivatedEventArgs args) override.

**Comment by Felicity-R** [comment]
> > Do you remember the API that made things "just work" before in Xamarin? It doesn't seem right to me to make Window.Activate also call SetForegroundWindow, but it could be I'm missing something.
> 
> It would have been UWP instead of WinUI3. We just called Window.Current.Activate(), from the Windows.UI.Xaml.Application.OnActivated(IActivatedEventArgs args) override.

Oh yeah, I just found the docs for that windows.ui.xaml (UWP) Activate method and it does explicitly say that it will bring it to the foreground:
> Attempts to activate the application window by bringing it to the foreground and setting the input focus to it.
https://learn.microsoft.com/en-us/uwp/api/windows.ui.xaml.window.activate?view=winrt-28000#windows-ui-xaml-window-activate

And then the microsoft.ui.xaml (WinUI3) Activate method does say the same thing:
> Attempts to activate the application window by bringing it to the foreground and setting the input focus to it.
https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.ui.xaml.window.activate?view=windows-app-sdk-2.0#microsoft-ui-xaml-window-activate

**Comment by JesseCol** [comment]
> > > Do you remember the API that made things "just work" before in Xamarin? It doesn't seem right to me to make Window.Activate also call SetForegroundWindow, but it could be I'm missing something.
> > 
> > 
> > It would have been UWP instead of WinUI3. We just called Window.Current.Activate(), from the Windows.UI.Xaml.Application.OnActivated(IActivatedEventArgs args) override.
> 
> Oh yeah, I just found the docs for that windows.ui.xaml (UWP) Activate method and it does explicitly say that it will bring it to the foreground:
> 
> > Attempts to activate the application window by bringing it to the foreground and setting the input focus to it.
> > https://learn.microsoft.com/en-us/uwp/api/windows.ui.xaml.window.activate?view=winrt-28000#windows-ui-xaml-window-activate
> 
> And then the microsoft.ui.xaml (WinUI3) Activate method does say the same thing:
> 
> > Attempts to activate the application window by bringing it to the foreground and setting the input focus to it.
> > https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.ui.xaml.window.activate?view=windows-app-sdk-2.0#microsoft-ui-xaml-window-activate

Thanks!  Yes, I can see now why this is a regression from UWP XAML -> WinUI3.

I have a PR open on fixing the WinUI3 doc, I'll try to get that in soon.

**Comment by riverar** [comment]
@JesseCol I think this issue can be closed. @Balkoth is confused about foreground rights and active windows vs. foreground windows. Maybe he could consult with someone with more experience. (Tongue in cheek.)

I don't think we should be promoting or supporting any workarounds that bypass foreground rights checks, to shield the user from misbehaving apps. (I believe there have been a few long-known holes.)

That all said, there may be a bug here with the single instancing architecture where the starting process has the shell's foreground rights and isn't passing them to the existing instance before terminating. (I haven't had a chance to look yet; It's certainly not helpful that there's no repro provided.)

**Comment by Balkoth** [comment]
> Here is the sample project: [WindowActivateTest.zip](https://github.com/microsoft/microsoft-ui-xaml/files/10220167/WindowActivateTest.zip)

Perhaps read the whole thread before jumping to conclusions.

**Comment by riverar** [comment]
@Balkoth Perhaps fill out the bug form correctly next time.

**Comment by riverar** [comment]
@JesseCol `DesktopWindowImpl::ActivateImpl` appears to be missing a call to `SetForegroundWindow` to take advantage of the foreground rights it may receive. This impl. only appears to call out to `ShowWindow`, `UpdateWindow`, and `SetActiveWindow`:

https://github.com/microsoft/microsoft-ui-xaml/blob/188f602b27cdb47572b28c380e9c087b02e1ccee/dxaml/xcp/dxaml/lib/DesktopWindowImpl.cpp#L403-L425

`ShowWindow` only sets the show state (maximized/restored/etc.), `UpdateWindow` only forces a repaint, and `SetActiveWindow` only updates the active window for the thread's msg queue. None of these grant system foreground focus.

The documentation on this API is also wrong, as it misleads the caller into thinking foreground attempt is made when one is never made.
> //     Attempts to activate the application window by bringing it to the foreground
> //     and setting the input focus to it.

Note WPF does this today:
https://github.com/dotnet/wpf/blob/main/src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/Window.cs#L483-L497

**Comment by riverar** [comment]
Oh looks like there's no appetite to change this API and instead there's a discussion about introducing a new foreground-setting API:
https://github.com/microsoft/microsoft-ui-xaml/pull/11240

**Comment by ChristianGalla** [comment]
@JesseCol 

Your proposal and this thread are discussing the current implementation in the different frameworks. But the focus should be that the WinUI3 API is easy to use, self explaining and does what the developer expects, ideally without having to look at the documentation. Introducing additional methods increases the complexity from developer perspective.

Is there any reason a Window activate call should not bring it to the foreground (if accepted by the OS)?
I don't know any reason from app developer perspective why an attempt to activate a Window should not set it to the foreground of all the Windows of the current app.
If the app is the last one, the user interacted with, I also don't see any reason why it should not set the Window to the global foreground of all apps.

Note: last interaction also means for example interactions with a toast notification of the app even if the app is not running during the toast notification click and / or if the app is single instance.

**Comment by hexawyz** [comment]
Yeah, to be honest, I don't understand what I've been reading on this issue recently. Is seems clear that the API is misleading from a consumer PoV, and the reason I have been following this issue for a long time is exactly because I have been it by the issue myself like most here. ([Here](https://github.com/hexawyz/Exo/blob/master/src/Exo/Ui/Exo.Settings.Ui/Program.cs#L71), the application won't be brought to the front when activated a second time)

The bug seems pretty straightforward: Practically everybody expects the function to hehave as the documentation says it should: bring that Window to the front. Therefore we consider that it should, and how that is implemented is of no concern to most of us, TBH.

The only question I have after reading the recent comments on the issue is in what situation on earth is the current behavior of `Activate()` useful to anybody in any way at all?

**Comment by JesseCol** [comment]
Thanks for the feedback!

I'm working on updating the doc page for Window.Activate() to fix it, hopefully that will be up soon.

Message received from y'all that the current behavior of Window.Activate() is surprising.

The tricky thing with changing Window.Activate() to call SetForegroundWindow is that WinUI apps have been using Window.Activate() as the "normal" "show-this-window" API for years.  I'm thinking it will be more confusing than clarifying to change its behavior at this point.  Ideally this API would probably have been called "Show", but unfortunately WinUI inherited the name "Activate" from UWP Xaml.  I agree it _is_ confusing that WPF's Activate() does something else.  But maybe there's a middle ground.  For example: we could deprecate Activate() and instead add Show() and TryActivate() -- is that more clear and intuitive?

I also want to make sure folks realize that calling SetForegroundWindow won't reliably give the window keyboard focus and bring it to foreground in all cases.  It merely _requests_ to Windows that it does.  If some other unrelated app has foreground, your window's taskbar icon will just flash.  The scenarios I'm aware of where this is needed are when another process is "handing-off" foreground rights to another process, using AllowSetForegroundWindow.  These are scenarios like single-instancing, activate-from-toast-notification, and deep-linking. 

If this issue is specifically about making window.Activate() bring the window to the foreground even in cases where SetForegroundWindow disallows it, that's a different topic, and I'll move this discussion about SetForegroundWindow to a more appropriate issue.

Thanks!

**Comment by riverar** [comment]
> The tricky thing with changing Window.Activate() to call SetForegroundWindow is that [...]

Fair enough.

> For example: we could deprecate Activate() and instead add Show() and TryActivate() -- is that more clear and intuitive?

Sounds OK to me, but we should be careful in naming--showing a window != activating a window != bringing a window to foreground.

Thanks for working on this.

**Comment by JesseCol** [comment]
> Sounds OK to me, but we should be careful in naming--showing a window != activating a window != bringing a window to foreground.

Agreed, which is why I actually kinda like having window.Activate() and window.TrySetForeground(), especially after we update the window.Activate() docs.  BUT, I do also get the appeal of matching WPF's APIs.

**Comment by Felicity-R** [comment]
> The tricky thing with changing Window.Activate() to call SetForegroundWindow is that WinUI apps have been using Window.Activate() as the "normal" "show-this-window" API for years. I'm thinking it will be more confusing than clarifying to change its behavior at this point. 

I don't really understand this. If the app is minimized, it already does come to the foreground when Activate is called, so foregrounding is not entirely new behaviour. And given that the documentation has said all along that it will foreground your app, anyone who is *relying* on it not foregrounding the app is basically relying on incorrect behaviour. I can't imagine a scenario where anyone would be upset with this change.

**Comment by Felicity-R** [comment]
> If some other unrelated app has foreground, your window's taskbar icon will just flash.

I just re-tested my own app; I had the app open, and then brought some Chrome windows in front of it, and when I used a deep link SetForegroundWindow did bring my app to the foreground just as I would like.

Edit: And if I remove SetForegroundWindow, it still comes to the foreground if either closed or minimized, just not when it's behind a different foreground window.

**Comment by JesseCol** [comment]
>I don't really understand this. If the app is minimized, it already does come to the foreground when Activate is called, so foregrounding is not entirely new behaviour. And given that the documentation has said all along that it will foreground your app, anyone who is _relying_ on it not foregrounding the app is basically relying on incorrect behaviour. I can't imagine a scenario where anyone would be upset with this change.

It's possible I'm being overly cautious, but it seems like a big change to make.  In WinUI, calling Window.Activate() is the main way you show a window -- it's what the templates do.  So if we had it also call SetForegroundWindow, we'd be making it so that this call happens every time an app shows a WinUI window -- I'm not aware of another framework that does this.  (lmk if this is more common than I realize)

Here's an example compat scenario: if an app creates/shows a secondary window while it's not in the foreground, after this change the user would see the taskbar icons for those windows flash.  (they didn't before).

>and when I used a deep link SetForegroundWindow did bring my app to the foreground just as I would like.

Great!  Hopefully we can help your scenario, then.  I suspect Windows does something to give your process/app foreground rights in this case, I admit I don't know exactly how this works.

**Comment by riverar** [comment]
We have to assume because of the bug's existence that someone is relying on it. So I think changing the API in this major release cycle is off the table.

**Comment by Felicity-R** [comment]
>  In WinUI, calling Window.Activate() is the main way you show a window

What do you mean by "show a Window"?
Let's say my app is running, and I open a new Chrome window (or any other app) and drag it in front of my app's window. Then I do something (click a launch link), which calls Window.Activate to "show" my app (and maybe do some other stuff in the app like navigate to a particular page). I would expect that to mean that now my app is visible (shown) in front of Chrome, i.e. foregrounded. Is that not what you mean by show? That is not what Window.Activate() does right now; my app remains hidden behind Chrome, and the taskbar icon does not even flash - it looks like nothing has happened.

If I minimize my app first, then it does get shown/foregrounded even if I was interacting with Chrome (Chrome is foregrounded). That inconsistency is what makes this feel buggy; if Windows.Activate didn't foregrounded the app in that scenario either, at least it would seem more consistent.

> Here's an example compat scenario: if an app creates/shows a secondary window while it's not in the foreground, after this change the user would see the taskbar icons for those windows flash. (they didn't before).

Doesn't that seem like a good thing? Usually if an app is trying to do something in the background, like trying to activate, or displaying some kind of alert, seeing the icon flash to get my attention is common behaviour.

**Comment by Felicity-R** [comment]
> We have to assume because of the bug's existence that someone is relying on it.

Would that not be true of every bug, then? 😜

Of course, since we have the workaround of calling SetForegroundWindow ourselves, the stakes are not super high here for those of us affected by the bug. I can understand it just not being worth the risk if there are too many unknowns about how people might be using it.

**Comment by riverar** [comment]
> Let's say my app is running, and I open a new Chrome window (or any other app) and drag it in front of my app's window. Then I do something (click a launch link), which calls Window.Activate to "show" my app (and maybe do some other stuff in the app like navigate to a particular page). I would expect that to mean that now my app is visible (shown) in front of Chrome, i.e. foregrounded. Is that not what you mean by show? That is not what Window.Activate() does right now; my app remains hidden behind Chrome, and the taskbar icon does not even flash - it looks like nothing has happened.

In this very specific activation scenario, everything appears to be working as currently designed (but not expected, understandably).

The browser window in the foreground is not interrupted by the subsequent launches because when your app calls Activate, WinUI 3 fails to take advantage of foreground rights given to it by failing to call `SetForegroundWindow`. In the minimized (iconic) state, WinUI 3's Activate implementation calls `ShowWindow(SW_RESTORE)` which restores the window. This is a legitimate user-facing transition and will bring the app to the foreground as a side effect.

So a future release of WinUI 3 should have a `TrySetForeground` method for you to put in your `Activated` handler. You should not, however, take away that calling `SetForegroundWindow` will always work--it's only working for you because your app has rights to do so. That will not always be the case.

**Comment by Felicity-R** [comment]
Ok, thanks. So 'show' means... just not minimized (but not necessarily visible/foregrounded, even when it does have the right to become so)?

> So a future release of WinUI 3 should have a `TrySetForeground` method for you to put in your `Activated` handler. You should not, however, take away that calling `SetForegroundWindow` will always work--it's only working for you because your app has rights to do so. That will not always be the case.

What's the point of a new TrySetForeground when we already have SetForegroundWindow? How will it be different?

Of course I do not expect my app to be foregrounded if it cannot get the rights to do so, that's not a problem.

**Comment by bogdan-patraucean** [comment]
> What's the point of a new TrySetForeground when we already have SetForegroundWindow? How will it be different?

@Felicity-R , `SetForegroundWindow ` is a native win32 API you need to access in a more complex way (like PInvoke, LibraryImport, etc.), while `TrySetForeground ` would be a method available on the `Window` class directly so you don't need to worry about finding a way to call the native API yourself, as it calls it for you internally.
